File: apps/api/momento_api/routes/timeline.py
from collections import defaultdict
from datetime import datetime
from typing import Annotated, Literal, Optional

from fastapi import APIRouter, Depends

from momento_api.auth.dependencies import CurrentUser, get_current_user
from momento_api.database import fetch_all
from momento_api.models.media import MediaResponse, TimelineGroup, TimelineListRequest, TimelineListResponse

router = APIRouter(prefix="/timeline", tags=["timeline"])


def _get_group_key(date_taken: Optional[str], group_by: Literal["year", "month", "week", "day"]) -> str:
    if not date_taken:
        return "Unknown"

    try:
        dt = datetime.fromisoformat(date_taken.replace("Z", "+00:00"))
    except (ValueError, AttributeError):
        return "Unknown"

    if group_by == "year":
        return str(dt.year)
    elif group_by == "month":
        return f"{dt.year}-{dt.month:02d}"
    elif group_by == "week":
        year, week, _ = dt.isocalendar()
        return f"{year}-W{week:02d}"
    else:
        return date_taken[:10]


def _row_to_media_response(row) -> MediaResponse:
    return MediaResponse(
        id=row["id"],
        filename=row["filename"],
        original_filename=row["original_filename"],
        media_type=row["media_type"],
        mime_type=row["mime_type"],
        width=row["width"],
        height=row["height"],
        file_size=row["file_size"],
        duration_seconds=row["duration_seconds"],
        date_taken=row["date_taken"],
        gps_latitude=row["gps_latitude"],
        gps_longitude=row["gps_longitude"],
        camera_make=row["camera_make"],
        camera_model=row["camera_model"],
        iso=row["iso"],
        exposure_time=row["exposure_time"],
        f_number=row["f_number"],
        focal_length=row["focal_length"],
        gps_altitude=row["gps_altitude"],
        location_state=row["location_state"],
        location_country=row["location_country"],
        keywords=row["keywords"],
        created_at=row["created_at"],
    )


def _fetch_timeline_rows(user_id: int, cursor: Optional[str], limit: int) -> list[dict]:
    base_query = "SELECT * FROM media WHERE user_id = ? AND deleted_at IS NULL"
    order_clause = " ORDER BY date_taken DESC, id DESC LIMIT ?"

    if cursor:
        cursor_parts = cursor.split("_")
        if len(cursor_parts) == 2:
            cursor_date, cursor_id = cursor_parts
            query = f"{base_query} AND (date_taken < ? OR (date_taken = ? AND id < ?)){order_clause}"
            return fetch_all(query, (user_id, cursor_date, cursor_date, int(cursor_id), limit + 1))

    # Default case (no cursor or invalid cursor structure basic fallback)
    query = f"{base_query}{order_clause}"
    return fetch_all(query, (user_id, limit + 1))


@router.post("/list", response_model=TimelineListResponse)
async def list_timeline(
    request: TimelineListRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> TimelineListResponse:
    # Cap limit to prevent large fetches
    limit = min(request.limit, 500)

    rows = _fetch_timeline_rows(current_user.id, request.cursor, limit)

    has_more = len(rows) > limit
    rows = rows[:limit]

    grouped: dict[str, list[MediaResponse]] = defaultdict(list)
    for row in rows:
        date_str = _get_group_key(row["date_taken"], request.group_by)
        grouped[date_str].append(_row_to_media_response(row))

    groups = [TimelineGroup(date=date, media=media) for date, media in grouped.items()]

    next_cursor = None
    if has_more and rows:
        last = rows[-1]
        next_cursor = f"{last['date_taken']}_{last['id']}"

    return TimelineListResponse(groups=groups, next_cursor=next_cursor, has_more=has_more)

File: apps/api/src/routes/timeline.rs
use axum::{extract::State, routing::post, Json, Router};
use chrono::{Datelike, NaiveDateTime};
use indexmap::IndexMap;

use crate::auth::{AppState, CurrentUser};
use crate::database::{fetch_all, queries};
use crate::error::{AppError, AppResult};
use crate::models::{MediaResponse, TimelineGroup, TimelineListRequest, TimelineListResponse};

pub fn router() -> Router<AppState> {
    Router::new().route("/timeline/list", post(list_timeline))
}

fn get_group_key(date_taken: Option<&str>, group_by: &str) -> String {
    let date_taken = match date_taken {
        Some(dt) => dt,
        None => return "Unknown".to_string(),
    };

    // Parse date - try ISO format first
    let dt = if let Ok(dt) = NaiveDateTime::parse_from_str(date_taken, "%Y-%m-%dT%H:%M:%S") {
        dt
    } else if let Ok(dt) = NaiveDateTime::parse_from_str(&date_taken.replace("Z", ""), "%Y-%m-%dT%H:%M:%S%.f") {
        dt
    } else if date_taken.len() >= 10 {
        // Just date
        if let Ok(d) = chrono::NaiveDate::parse_from_str(&date_taken[..10], "%Y-%m-%d") {
            d.and_hms_opt(0, 0, 0).unwrap()
        } else {
            return "Unknown".to_string();
        }
    } else {
        return "Unknown".to_string();
    };

    match group_by {
        "year" => dt.year().to_string(),
        "month" => format!("{}-{:02}", dt.year(), dt.month()),
        "week" => {
            let week = dt.iso_week();
            format!("{}-W{:02}", week.year(), week.week())
        }
        _ => date_taken.chars().take(10).collect(),
    }
}

async fn list_timeline(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<TimelineListRequest>,
) -> AppResult<Json<TimelineListResponse>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;
    let limit = request.limit.min(500);

    let rows = if let Some(ref cursor) = request.cursor {
        let parts: Vec<&str> = cursor.split('_').collect();
        if parts.len() == 2 {
            let cursor_date = parts[0];
            let cursor_id: i64 = parts[1].parse().unwrap_or(0);
            fetch_all(
                &conn,
                queries::timeline::SELECT_PAGINATED,
                &[&current_user.id, &cursor_date, &cursor_date, &cursor_id, &(limit + 1)],
                map_timeline_row,
            )?
        } else {
            fetch_default_timeline(&conn, current_user.id, limit)?
        }
    } else {
        fetch_default_timeline(&conn, current_user.id, limit)?
    };

    let has_more = rows.len() > limit as usize;
    let rows: Vec<_> = rows.into_iter().take(limit as usize).collect();

    // Group by date using IndexMap to preserve insertion order
    // (rows are already ordered by date_taken DESC from database)
    let mut grouped: IndexMap<String, Vec<MediaResponse>> = IndexMap::new();
    for (media, date_taken) in &rows {
        let key = get_group_key(date_taken.as_deref(), &request.group_by);
        grouped.entry(key).or_default().push(media.clone());
    }

    let groups: Vec<TimelineGroup> = grouped
        .into_iter()
        .map(|(date, media)| TimelineGroup { date, media })
        .collect();

    let next_cursor = if has_more && !rows.is_empty() {
        let (last, last_date) = rows.last().unwrap();
        last_date.as_ref().map(|dt| format!("{}_{}", dt, last.id))
    } else {
        None
    };

    Ok(Json(TimelineListResponse {
        groups,
        next_cursor,
        has_more,
    }))
}

fn fetch_default_timeline(
    conn: &crate::database::DbConn,
    user_id: i64,
    limit: i32,
) -> AppResult<Vec<(MediaResponse, Option<String>)>> {
    fetch_all(
        conn,
        queries::timeline::SELECT_DEFAULT,
        &[&user_id, &(limit + 1)],
        map_timeline_row,
    )
}

fn map_timeline_row(row: &rusqlite::Row) -> rusqlite::Result<(MediaResponse, Option<String>)> {
    let date_taken: Option<String> = row.get(9)?;
    let media = MediaResponse {
        id: row.get(0)?,
        filename: row.get(1)?,
        original_filename: row.get(2)?,
        media_type: row.get(3)?,
        mime_type: row.get(4)?,
        width: row.get(5)?,
        height: row.get(6)?,
        file_size: row.get(7)?,
        duration_seconds: row.get(8)?,
        date_taken: date_taken.clone(),
        gps_latitude: row.get(10)?,
        gps_longitude: row.get(11)?,
        camera_make: row.get(12)?,
        camera_model: row.get(13)?,
        lens_make: row.get(14)?,
        lens_model: row.get(15)?,
        iso: row.get(16)?,
        exposure_time: row.get(17)?,
        f_number: row.get(18)?,
        focal_length: row.get(19)?,
        focal_length_35mm: row.get(20)?,
        gps_altitude: row.get(21)?,
        location_city: row.get(22)?,
        location_state: row.get(23)?,
        location_country: row.get(24)?,
        video_codec: row.get(25)?,
        keywords: row.get(26)?,
        created_at: row.get(27)?,
    };
    Ok((media, date_taken))
}
